lib/ash_json_api/resource/transformers/require_primary_key.ex

```diff
@@ -0,0 +1,37 @@
+defmodule AshJsonApi.Resource.Transformers.RequirePrimaryKey do
+  @moduledoc "Ensures that the resource either has a primary key 
+    or includes primary_key section from json_api if it has a composite key
+  "
+
+  use Ash.Dsl.Transformer
+  alias Ash.Dsl.Transformer
+
+  def transform(_resource, dsl) do
+    case Transformer.get_option(dsl, [:json_api, :primary_key], :keys) do
+      nil ->
+        dsl
```

Comment:
You can use `Ash.Resource.primary_key/2` here, if you add a clause to ensure that this check goes after the check that validates/caches primary keys (probably a good idea anyway, so they get a sensible error first):

```elixir
# At the bottom of the module
def after(Ash.Resource.Transformers.CachePrimaryKey), do: true
def after(_), do: false
```

Reply:
Then you can say 
```
case Ash.Resource.primary_key(resource) do
  [_] -> {:ok, dsl}
  _ -> ...
end
```

Reply:
Added `&after?/1`